### PR TITLE
Prepare the `bootstrap` tool for the new check-cfg syntax

### PR DESCRIPTION
**Pull request by Urgau**
This PR prepare the `bootstrap` tool for the [new check-cfg syntax](https://github.com/rust-lang/rust/pull/111072) as well as the according [changes to Cargo](https://github.com/rust-lang/cargo/pull/12845).

~~Note that while the new syntax can technically available on stage > 2, we actually cannot use it since we need a cargo version that supports the new syntax which won't happen until the next beta bump (if I understand everything correctly).~~

r? bootstrap

### CODE REVIEW

**Comment by weihanglo** [comment]
Hit this issue <https://github.com/rust-lang/rust/pull/117000#issuecomment-1773657187>. How can we help this moving forward?

**Comment by Urgau** [comment]
> Hit this issue [#117000 (comment)](https://github.com/rust-lang/rust/pull/117000#issuecomment-1773657187). How can we help this moving forward?

This PR shouldn't be necessary until the next beta bump. Something unexpected is going on.

**Comment by Urgau** [comment]
I've added a commit to enable the new `{-Z,--}check-cfg` arguments when doing a "local rebuild" (build the same version of rustc with it-self) and also added a revert commit of https://github.com/rust-lang/rust/pull/117018 since it's no longer necessary.

File: src/bootstrap/src/bin/rustdoc.rs [hunk 1]
```diff
@@ -70,7 +70,9 @@ fn main() {
         cmd.arg("--cfg=bootstrap");
     }
     cmd.arg("-Zunstable-options");
+    // #[cfg(bootstrap)]
     cmd.arg("--check-cfg=values(bootstrap)");
```
Comment (Kobzol) [inline src/bootstrap/src/bin/rustdoc.rs:74]: I wonder if this should be also gated for local rebuilds, but I don't think we can even set a custom rustdoc in `config.toml`, so it probably doesn't matter.

**Review by Kobzol — COMMENTED** [review]
I tried a few configurations locally and they were all working.

Hopefully, this will fix bootstrap timings on perf.RLO, so we should do that as soon as possible. Let's merge it and see if anything else breaks.

**Comment by lqd** [comment]
image

🎉

**Pull request merged**